### ISSUE TITLE
Add chunkloading active chunk to pumps, miners, fluid drills, and ore drills

### ISSUE THREAD

**Issue by richardhendricks** [GTNewHorizons/NewHorizons#4546]

**Comment by Ngar1** [comment]
Using an anchor is a solution and has also a time limit. 
My be a solar powered chunkloading pump with endless water (may ways) could be used as endless chunkloader

**Comment by richardhendricks** [comment]
So, basic design document:

Machines will only chunkload their active chunk, not themselves (well, until the active chunk is the chunk they are in anyways.).

To keep the machine controller chunkloaded will require a normal anchor.  This way servers can enforce their normal chunkloading rules. The Railcraft anchors support an Anchor Sentinel that allows for reducing the area of effect of the anchor to 1 by X. For these machines the anchor could use only 1 chunk, so the time left would be 9x more.

This will also require changing machines so they behave better wrt to how they do their work, ie instead of doing X by Z by Y, the machine will need to at least do a X by Z in a chunk before doing the next chunk.  Maybe doing the whole chunk by Y before the next chunk, but I think that probably isn't necessary.  Will need to check on drilling speed, if an X by Z is too fast - say less than a minute - probably should just do the whole chunk before going to the next chunk to prevent chunk churn..

For fluid pumps it is a little different.  They currently pump from the outside - in based on distance from the pumphead. Since it is distance oriented, initially they pump the furthest corners in a round-robin but after that, the withdrawal area is circular around the pump. It would be difficult to pump everything in one X by Z in a chunk before progressing to the next chunk. Might have to add a 2nd queue that has fluid blocks sorted by chunk or chunk distance, and choose which fluid block to pull based on that.  Will need to check on actual behaviour of pumps - do they pull in a circle around the pump that ie rotates in a specific direction as it goes, or does it jump all over the place around the equidistant circle. 

Maybe setup the pumping queue is such a way that we include distance in the queue but massaged so that the chunk number is part of the hashing, ie truncate distance to distance >> 2 or >> 4 but then append a chunk-unique number and sort the pump queue by that.  So nearly-equidistant blocks in a chunk are naturally grouped together and pumped together. Fluid pumps aren't super fast, but grouping 30-64 blocks together in a chunk is probably a good idea.

**Comment by observeroftime02** [comment]
I suggest the notion that for Ore Drilling Rigs, whole chunks should be completed before moving on to the next chunk. The tier 4 drilling rig, for example, when powered by an UHV hatch, moves so fast that chunk churning would be inevitable. Currently it works on a layer by layer basis, not a chunk by chunk one. So, depending on how much ore is in a given layer, it will speed through chunks too quickly. A given layer in a 9 x 9 area might only have a few hundred ores in it, and then it will rapidly zip through the different chunks before moving on to the next layer. 

So, IMHO it only makes sense to complete a single chunk before mining the next one. It would also improve (depending on how you look at it) the mechanics of the drilling rig, because it would be able to mine different ore layers faster. If I set up a drilling rig, I would have to wait a long time before it reaches the deeper ore layers, because everything on top would have to be removed first. That might mean that my drilling rig never gets to the deeper layers before power / drilling fluid runs out. Mining on a chunk by chunk basis would mean I get the deeper layer ores faster, and get a better return initially when setting up a drilling rig. In the end though, it would still mine the same when running it to completion. I'm therefore in favor of doing it chunk by chunk, because it means returning useful ores faster (although in smaller amounts initially).

**Comment by richardhendricks** [comment]
I would have to set it up to drill all the way to the bedrock first before removing ores. It would add a little complexity to the operation.

**Comment by observeroftime02** [comment]
Sure, but what's the alternative? Any ideas? Perhaps the drill can go up and down again for each chunk. Otherwise a whole spread of 9x9 would need to be chunkloaded due to the speed of the drilling rig, or continuous rapid loading and unloading of chunks would occur, which would be undesirable for performance. Ideally no chunk would need to be loaded and unloaded more than once. That would be the smallest performance impact. 

Just some ideas.

**Comment by richardhendricks** [comment]
Yeah it's probably the most logical. We are looking at a speed nerf for the ore drilling rigs too now that they have the fortune boost.

**Comment by repo-alt** [comment]
Any progress in that area?

**Comment by repo-alt** [comment]
@richardhendricks If you won't have time for that issue on vacation, I might take a look at it, the functionality seems to be rather important, especially for high tier drills.

**Comment by richardhendricks** [comment]
I have a half started branch at https://github.com/GTNewHorizons/GT5-Unofficial/tree/Chunkload_dev

**Comment by mitchej123** [comment]
Oh nice, I might also take a look at this; didn't want to start from scratch.  @repo-alt happy to pow-wow on this if you want

**Comment by repo-alt** [comment]
Eh @mitchej123, it depends if you have much free time in the nearest future or not, or we might separate the work, because there are at least 3 parts to be done: Chunkload manager (already started by @richardhendricks ), drill refactoring (reordering digging by chunk), pump refactoring (reordering fluid blocks queue by chunk)

**Comment by richardhendricks** [comment]
For a test I was going to try a dumb implementation that requests tickets for wvery block, and have the chunkload manager check against a list of ticketed chunks. Then later make the chunk requests smarter

**Comment by repo-alt** [comment]
As far as I understand, currently ore drill digs layer by layer, so, requesting every block will load entire area in no time. If digging is reordered, then depending on _how_ it is reordered, it may be easier or not to request loading for every block, but it does not solve release problem, because surely we don't want to keep processed chunks loaded until server restart?

**Comment by richardhendricks** [comment]
.It was just for testing lol to see if tickets even load chunks etc, not as the released code

**Comment by repo-alt** [comment]
@richardhendricks 
I'm not sure I like the idea of IChunkLoader interface on the tile entity. As I understand it was intended more like a disguised mixin/trait base class, but in Java interfaces all data members are constant, no way you would be able to assign that machineTicket member after construction. All in all, GT_ChunkManager looks like good enough place for all related functionality. It may keep some multimaps and such, for bookkeeping. I looked at the Railcraft and ExU chunkloading functionality and in general Forge chunkloading interface is usable/reasonable enough.
@mitchej123 
I will then start with making some stub chunkloading implementation. You may join the work if you like, with doing pump or miner part.

**Comment by repo-alt** [comment]
Ok I did some stub chunkloading, at least for debugging, looking at the miner, and I have a question/RFC:
Miners and Drills use mining pipes, so, how should the new behaviour look - should the miner once lower the mining pipes in the central chunk, and work in all the chunks with the head at the bedrock, or should it retract/lower the pipes for each chunk? Also what to do if there are not enough pipes, process so much layers in each chunk how much pipes are provided? What if the pipes are added at later stage, when the first chunk is already processed?

**Comment by botn365** [comment]
i remeber it would be changed so it first lowers the pipes until it cant and then mine a  chunk until it is empty and move to the next

**Comment by repo-alt** [comment]
Remember? Where it was discussed?

**Comment by repo-alt** [comment]
Another consideration: what if the player does not want to mine all the area, and just wants this and that veins? Now he can place a couple of chunkloaders on vein centers and achieve exactly that. So, should the chunk loading be optional? If yes - how?

**Comment by richardhendricks** [comment]
meh, I think that usage model is pretty small. If a player wants to target specific veins, they can use the singleblock miners IMO.

**Comment by repo-alt** [comment]
Well, I added turning off chunkloading capability with Soldering iron (Bart usurped the Screwdriver), it is not that complicated.

**Comment by repo-alt** [comment]
As far as I can see Oil drill does not need chunkloading, it reads/writes information from the chunk directly. Singe block miner is a low priority because it's radius is small, (and who uses EV single block miner, seriously) So, the only machine to look at is a pump. I read the second comment above, and still I don't quite see how desired behaviour should look.

**Comment by repo-alt** [comment]
Looked at the pump code and found out it has chunkloading capability as is. It is not "real" chunkloading, but it (probably) works anyway. Actually, in GT special efforts are taken to _prevent_ chunkloading of remote tiles. There is a variable "ignoreUnloadedChunks" in BaseTileEntity and functions like getBlock, etc. explicitly check tor it, and return air block for unloaded chunks. Pump is the only machine that unsets this variable, and does fetch blocks from unloaded chunks. Need to test that just in case, but if it does not cause much lag as it is I would leave it alone.

**Comment by richardhendricks** [comment]
Sounds great.  The issue is that the chunk load/unload because it is a distance-sorted queue would skip around the chunks in a circle. It's only a few chunks at a time and how often isn't super fast, so it might not be a *big* deal.

My original plan was to sort the blocks from the current distance-only queue into a new queue that is sorted by chunk, with blocks in the same chunk kept together. But the algo would need to be really tweaked to accomplish that.

**Comment by richardhendricks** [comment]
Hmmm...One reason the pump skips is because it does a square area. Once it picks the corners out, it should be doing a circle that stays in a chunk for a couple of blocks before proceeding to the next chunk.

**Comment by repo-alt** [comment]
well, I can easily imagine situations when some blocks will be inaccessible when doing chunk-by chunk traversal in one pass

**Comment by richardhendricks** [comment]
Yeah, we're probably over-optimizing this anyways. 

How about this?  Just have the pump request a ticket for each chunk it is pumping in. Don't worry about releasing the tickets. On a server restart, don't restore the tickets either. At least that way we won't be churning the chunks while the pump is active.

**Comment by repo-alt** [comment]
It will load all the pumping area while building a queue, for ZPM pump this is quite an area. That would probably be more load for the server to keep all that area loaded

**Comment by richardhendricks** [comment]
Good point. :-/ Does anyone actually use more than an HV pump? Maybe we should just disable them above EV.

**Comment by repo-alt** [comment]
Perhaps above HV it should be a magic multiblock working by a miner logic, fetching all the blocks regardless of accessibility.

**Comment by BlackFlameTNT** [comment]
I would say ppl might use EV+ and probably even ZPM for collecting unknown water from T8 plants for seaweeds..... So this is the only thing being useful unless we add unknown water as oil.

**Comment by repo-alt** [comment]
Perhaps. There is also Ender Quarry pump upgrade with unlimited range

**Comment by richardhendricks** [comment]
I think for now we can leave the pump logic as-is, and file a new ticket if we decide we need to. Other than lava in the Nether, there really isn't too much need for large-area pumping right?
